### PR TITLE
Add post on What is Production Grade?

### PR DESCRIPTION
**Pull request by the-maldridge**
@phy1729 @m-wynn @dinoocch what do you think?

### CODE REVIEW

File: content/post/2019-01-20-what-is-production-grade.markdown [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+title: "What Is Production Grade?"
+date: 2019-01-20T15:35:03-08:00
+---
+
+I recently decided to try the Caddy webserver project's implementation
+of a webserver written in Golang.  It had many interesting things
+going for it, including automagic certificate renewal via ACME and
+Let's Encrypt, "plugins" written in a sane language, dramatically
+simplified configuration, and perhaps in one of the most boastful
+statements I've seen of any software recently, it claims to be
```
Comment (phy1729) [inline content/post/2019-01-20-what-is-production-grade.markdown:11]: English pedant hat: When a list contains commas, the list separator is promoted to a semicolon.

File: content/post/2019-01-20-what-is-production-grade.markdown [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+Ansible, I really question this claim.
+
+So that's what this article is about.  I'm putting down here what I
+think production grade software should be, and some of the problems
+with what software is out there.
+
+## What is Production?
+
+To answer the original question, we have to first look into what is
+production.  In the general sense, production is agreed to be the
```
Comment (phy1729) [inline content/post/2019-01-20-what-is-production-grade.markdown]: Wikipedia pedant hat: "production is agreed" `{{By whom|date=January 2019}}`

File: content/post/2019-01-20-what-is-production-grade.markdown [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+environment of a software system that is either directly or
+transitively involved in serving end-users.  These end-users can be
+internal or external to an organization, but the defining
+characteristic that makes someone an end user is that they can be
+unhappy about the status of a system and their unhappiness directly
+translates into a need to act for some team responsible for the
+system.  This is in contrast to dev or QA environments where
+unhappiness about the system state does not demand action.
```
Comment (phy1729) [inline content/post/2019-01-20-what-is-production-grade.markdown]: I'd add immediate before action. A dev environment broken for months isn't tenable.

File: content/post/2019-01-20-what-is-production-grade.markdown [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+
+Put more simply and more directly: if it is important enough to wake
+some one up and put a keyboard in front of them, its part of
```
Comment (phy1729) [inline content/post/2019-01-20-what-is-production-grade.markdown]: it's

File: content/post/2019-01-20-what-is-production-grade.markdown [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+production.
+
+With a reasonable definition of what production is, lets now look at
+what production grade means.
+
+## What is Production Grade?
+
+Like lumber, meat, and equipment, its not unreasonable to say that
```
Comment (phy1729) [inline content/post/2019-01-20-what-is-production-grade.markdown]: it's

File: content/post/2019-01-20-what-is-production-grade.markdown [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+software comes in various grades.  In many cases these terms are
+legally protected.  For example the United States Department of
+Agriculture governs the use of specific terms for grades of meat.
+Similarly the Western Wood Products Association provides grading
+criteria for lumber.  It is interesting then that no such set of
+standards exists for software.  Anyone can claim at any time that any
+software is production grade, or is of sufficient quality to wake
+someone up at night.
+
+Its at this point that I'm going to clearly state that for the
```
Comment (phy1729) [inline content/post/2019-01-20-what-is-production-grade.markdown]: It's

File: content/post/2019-01-20-what-is-production-grade.markdown [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+purposes of this article, I'm glossing over the complexity of software
+fitness for a particular purpose, such as life or safety critical
+systems.  These systems usually require the review and certification
+of a Professional Engineer, whose title and license are part of a
+regulated profession in most parts of the world.  I am not a
+Professional Engineer, and so I can't speak to the kinds of things
+that one would look for in certifying software.  I can only speak to
+what I look for in things that I allow to wake me up at night.
```
Comment (phy1729) [inline content/post/2019-01-20-what-is-production-grade.markdown]: The use of `things` in a meaning distinct from the use in the prior sentence is slightly confusing. Perhaps use `software` here instead?

File: content/post/2019-01-20-what-is-production-grade.markdown [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+
+Like most things, this is a sliding scale.  Lets start at the huge end
```
Comment (phy1729) [inline content/post/2019-01-20-what-is-production-grade.markdown]: Let's

File: content/post/2019-01-20-what-is-production-grade.markdown [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+grade (whether or not this was true is a debate for another time).
+
+In this environment, I personally held the definition that production
+grade wasn't something I could write down.  It was something I knew
+when I saw it, and was greater than the sum of its parts.  Those
+familiar with US Supreme Court decisions might find this equivalently
+infuriating to Justice Potter Stewart's test for obscenity.  When a
+developer wanted me to OK a feature to push to the many many tasks we
+had running around the world I had a checklist for certain, but I also
+had to ask myself what my gutt feeling was on the change.  Was it
```
Comment (phy1729) [inline content/post/2019-01-20-what-is-production-grade.markdown]: gut

File: content/post/2019-01-20-what-is-production-grade.markdown [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+safe?  Would it cause downtime?  How might this affect disaster
+response if we needed to shut off the new widget?  All of these are
+important questions to ask when looking for "Production Grade" at
+insane scales.
+
+Some of the biggest things I looked for had to do with startup and
+shutdown.  I really wanted to see that in anything I was expected to
```
Comment (phy1729) [inline content/post/2019-01-20-what-is-production-grade.markdown]: I'd move `in anything I was expected to be responsible for` to the start of the sentence.

File: content/post/2019-01-20-what-is-production-grade.markdown [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+obvious that something was wrong, as would an entire pool sitting in
+unhealthy state.
+
+Another really big thing I always looked for had to do with upgrades.
+The question to always ask here is what happens at the balance point
+when 50% of the fleet is on version N and 50% is on N+1.  In the best
+case nothing happens, in the worst case you can get persisted data
+corruption.
+
+These are all really important things to look at in defining the
```
Comment (phy1729) [inline content/post/2019-01-20-what-is-production-grade.markdown]: This is the fourth use of really in this section.

File: content/post/2019-01-20-what-is-production-grade.markdown [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+questions.
+
+### Moderate Scale
+
+For me moderate scale is the size of a medium sized company.  While at
+The University of Texas at Dallas, I was responsible for a computer
+network serving roughly 1400 people on a daily basis in both local
+computer lab format, and in remote terminal services format.
+
+In this environment, we built a lot of our own operations software,
```
Comment (phy1729) [inline content/post/2019-01-20-what-is-production-grade.markdown]: Who is we?

File: content/post/2019-01-20-what-is-production-grade.markdown [hunk 13]
```diff
@@ -0,0 +1,270 @@
...
+but the software we ran to provide services was almost all off the
+shelf.  We generally looked for a few specific features when selecting
+software.
+
+First and foremost it had to be maintained.  This seems obvious, but
+I've seen a lot of places get into terribly deep trouble over the
+years by hinging some critical process on legacy infrastructure that
+couldn't be kept up.  As a direct extension of this, its really
```
Comment (phy1729) [inline content/post/2019-01-20-what-is-production-grade.markdown]: it's

Comment (phy1729) [inline content/post/2019-01-20-what-is-production-grade.markdown]: Before we are considering software which never gets updates. This sentence goes to the opposite end of the spectrum with software which gets updates too frequently. Perhaps a phrase more oppositional than "as a direct extension" is appropriate.

File: content/post/2019-01-20-what-is-production-grade.markdown [hunk 14]
```diff
@@ -0,0 +1,270 @@
...
+important that there's a path between versions.  Breaking changes at
+major releases are all good and well, but if there's a major release
+every few months, that's a good sign that the project isn't stable,
+and therefore isn't ready to be run in production yet.
+
+One of the other big things that I'm looking for at this scale is how
+to get redundancy.  I need some way to stand up multiple copies of a
+service so that I can upgrade them one at a time, and so that if I
```
Comment (phy1729) [inline content/post/2019-01-20-what-is-production-grade.markdown]: I'd drop the second so that and swap the phrase to be
and scale out if I have more requests than a since instance can serve.

File: content/post/2019-01-20-what-is-production-grade.markdown [hunk 15]
```diff
@@ -0,0 +1,270 @@
...
+I also look at software at this scale to be debuggable.  For me this
+is one thing that Caddy didn't have.  I lost several hours trying to
+figure out what was wrong as the server would just hang on startup.  I
+eventually was able to isolate the fault to a single domain for a
+single virtualhost.  Several more hours of debugging isolated this to
+a lock that had been set in Caddy's state directory which needed to be
+manually cleared.  This could have been reduced to a 10 minute
+troubleshooting session if only Caddy had printed out that there was a
+lock being held during startup.  Instead, I had fumble my way through
+the startup process until I realized it had a lock.  Why these locks
```
Comment (phy1729) [inline content/post/2019-01-20-what-is-production-grade.markdown]: The rest of this paragraph feels like a footnote.

File: content/post/2019-01-20-what-is-production-grade.markdown [hunk 16]
```diff
@@ -0,0 +1,270 @@
...
+
+
+## To the Other Side
+
+"But wait!" I hear you say to your monitor, "I write software, not run
+other people's software like you!  How do I know that what I've
+written is ready to hand off to an operations team to deploy it and
+run it!?"
+
+This is a question I don't have an answer to.
```
Comment (phy1729) [inline content/post/2019-01-20-what-is-production-grade.markdown]: English pedant hat: to which I don't have an answer

File: content/post/2019-01-20-what-is-production-grade.markdown [hunk 17]
```diff
@@ -0,0 +1,270 @@
...
+
+For over a year now I've been working on a project called NetAuth.
```
Comment (phy1729) [inline content/post/2019-01-20-what-is-production-grade.markdown]: English pedant hat: comma after now.

File: content/post/2019-01-20-what-is-production-grade.markdown [hunk 18]
```diff
@@ -0,0 +1,270 @@
...
+Right now it satisfies all the requirements I have for running in the
+small to medium production system, but I'm still uncomfortable
+deploying it in prod.  Partially because it hasn't satisfied the
+many-eyes component that makes open source security softare good, but
+partially because I'm just not confident enough in my own software to
+say "yes, this is good, you should run it".
```
Comment (phy1729) [inline content/post/2019-01-20-what-is-production-grade.markdown:212]: English pedant hat: second comma should be bumped to a semicolon and the period goes inside the quotes.

File: content/post/2019-01-20-what-is-production-grade.markdown [hunk 19]
```diff
@@ -0,0 +1,270 @@
...
+
+Personally if you ever find an engineer that has that kind of attitude
```
Comment (phy1729) [inline content/post/2019-01-20-what-is-production-grade.markdown]: "that kind of attitude" is ambiguous. Your attitude or the attitude of the quote? (Obvious after reading the sentence, but not clear as reading the phrase.

File: content/post/2019-01-20-what-is-production-grade.markdown [hunk 20]
```diff
@@ -0,0 +1,270 @@
...
+towards softare you should think very carefully about the claims they
+make, and find out who's head is on a platter if they're wrong.
```
Comment (phy1729) [inline content/post/2019-01-20-what-is-production-grade.markdown]: whose

File: content/post/2019-01-20-what-is-production-grade.markdown [hunk 21]
```diff
@@ -0,0 +1,270 @@
...
+
+Interestingly, I would have no problems running Void Linux in prod.
+I've actually done this in the past, and my reasoning from then stands
+now.  I believe firmly that it is better to run a less battle hardened
+system that you can fit in entirely in your head rather than an
```
Comment (phy1729) [inline content/post/2019-01-20-what-is-production-grade.markdown:222]: Preach!

File: content/post/2019-01-20-what-is-production-grade.markdown [hunk 22]
```diff
@@ -0,0 +1,270 @@
...
+"industry standard" one that you can't.  Anyone who's had to
+troubleshoot a large complex system during an outage will know exactly
+why this is the case.
+
+
+## TL;DR
+
+I think posts like this always need a takeaway that you can think on
+later.  To distill down all the ideas that are in this post into
```
Comment (phy1729) [inline content/post/2019-01-20-what-is-production-grade.markdown]: "To distill down all the ideas that are in this post into something you can chew on later" doesn't add anything.

File: content/post/2019-01-20-what-is-production-grade.markdown [hunk 23]
```diff
@@ -0,0 +1,270 @@
...
+something you can chew on later, here's what I think is important in
+production software:
+
+  * Good logging.  You can always throw away log entries later, but
+    getting entries that the software never wrote is impossible.
+  * Instrumentation.  Like logging, production grade software was
+    designed with the intention that it might need to be looked at,
+    and so it should export metrics that tell you what's working, or
+    more importantly what isn't.
+  * Good Documentation.  While I love talking to people on IRC, I
```
Comment (phy1729) [inline content/post/2019-01-20-what-is-production-grade.markdown]: Lowercase D.

File: content/post/2019-01-20-what-is-production-grade.markdown [hunk 24]
```diff
@@ -0,0 +1,270 @@
...
+    prefer if the problems that I might face are well documented, and
```
Comment (phy1729) [inline content/post/2019-01-20-what-is-production-grade.markdown:240]: I'd change "I prefer if the problems that I might face are well documented, and are easy to find." to "I would prefer the problems that I might face be well documented in an easy to find place." unless you prefer the problems be easy to find and the documentation difficult :slightly_smiling_face:. If you keep the current structure, no comma here.

File: content/post/2019-01-20-what-is-production-grade.markdown [hunk 25]
```diff
@@ -0,0 +1,270 @@
...
+    are easy to find.  If the documentaiton is "Read the Source"
```
Comment (phy1729) [inline content/post/2019-01-20-what-is-production-grade.markdown]: Comma inside the quotes.

File: content/post/2019-01-20-what-is-production-grade.markdown [hunk 26]
```diff
@@ -0,0 +1,270 @@
...
+    that's not the end of the world, but the source had better be
```
Comment (phy1729) [inline content/post/2019-01-20-what-is-production-grade.markdown]: And then this comma is bumped to a semicolon.

File: content/post/2019-01-20-what-is-production-grade.markdown [hunk 27]
```diff
@@ -0,0 +1,270 @@
...
+Some of the more ethereal qualities revolve around the development
+process.  I might look into how many open issues there are, when they
+were last touched, and if it looks like the team is open to criticism
+towards problems in the software they've written (consequently this is
+one of the many reasons I don't believe systemd is production grade).
+
+If you're writing software and want to know if its production grade,
+the easiest way to find out is to go ask an operations team member.
+Either in your own organization or on the internet.  Most of us are
+more than happy to take a look at software if its interesting, solves
```
Comment (phy1729) [inline content/post/2019-01-20-what-is-production-grade.markdown:255]: it's

File: static/css/main.css [hunk 28]
```diff
@@ -47,6 +47,7 @@ h4 {
 
 ul {
     list-style: disc inside;
+    margin-bottom: 1em;
```
Comment (phy1729) [inline static/css/main.css:50]: Git pedant hat: this belongs in its own commit.

File: content/post/2019-01-20-what-is-production-grade.markdown [hunk 29]
```diff
@@ -0,0 +1,270 @@
...
+assumptions about how much of the rest of the world around it is
+functional.
+
+### Subminiature
+
+Choosing software that runs on my home network is perhaps the most
+picky I can be.
+
+In many cases, this is software running without redundancy on
+non-redundant hardware.  Add to this that its running in what most
```
Comment (phy1729) [inline content/post/2019-01-20-what-is-production-grade.markdown:175]: it's

File: content/post/2019-01-20-what-is-production-grade.markdown [hunk 30]
```diff
@@ -0,0 +1,270 @@
...
+people would consider to be a "privileged" environment and its easy to
```
Comment (phy1729) [inline content/post/2019-01-20-what-is-production-grade.markdown]: it's

File: content/post/2019-01-20-what-is-production-grade.markdown [hunk 31]
```diff
@@ -0,0 +1,270 @@
...
+    incredibly well commented in that case.
+
+
+Some of the more ethereal qualities revolve around the development
+process.  I might look into how many open issues there are, when they
+were last touched, and if it looks like the team is open to criticism
+towards problems in the software they've written (consequently this is
+one of the many reasons I don't believe systemd is production grade).
+
+If you're writing software and want to know if its production grade,
```
Comment (phy1729) [inline content/post/2019-01-20-what-is-production-grade.markdown]: it's

**Comment by phy1729** [comment]
As for what I think, I quite agree and certainly think people should prefer the ability to keep a system in one's head. Personally I also used the criterion "How easily can I hand this over?" since I knew the day would come.

File: content/post/2019-01-20-what-is-production-grade.markdown (hunk 4 above)
Comment (m-wynn) [reply content/post/2019-01-20-what-is-production-grade.markdown]: IDK, I have production apps with VERY loose SLAs
